### ISSUE TITLE
Packet c3.medium.x86 instances can't find root device because of ahci module failure

### ISSUE THREAD

**Issue by pothos** [flatcar/Flatcar#155]
**Description**

While the first boot works fine after the installation, following reboots get stuck in the initramfs because the root partition cannot be found. Reloading the `ahci` kernel module resolves this.

**Impact**

Flatcar Container Linux is not working on these instances (and maybe others).

**Environment and steps to reproduce**

1. **Set-up**: Select Flatcar for a c3.medium instance on Packet
2. **Task**: Reboot
3. **Action(s)**: Login via the OOB console and wait for the dracut shell after some minutes (up to 15) and press Enter (only 5 minutes time) to enter the shell which shows that `/dev/disk-by-label` is empty
  a. To resolve this, unload and load the `ahci` module, check that `/dev/disk-by-label` has the root partition, then exit to continue the boot
4. **Error**: This seems to be the related error message: `ata10.00: both IDENTIFYs aborted, assuming NODEV`

**Expected behavior**
The module works so that the root partition can be found

**Additional information**
@marga-kinvolk Did I miss something?

**Comment by pothos** [comment]
`m2.xlarge.x86` is also affected

**Comment by margamanterola** [comment]
While investigating this last month, I found this old issue that sounds very similar: https://sourceforge.net/p/scst/mailman/message/29423762/, the issue is for a different driver and has been fixed for a long time, but it's interesting that it's the same error and same behavior. In that case, the root cause was a missing zero initialization in the module.  It's very possible that this is also the case here.

Also, we've tested this with different versions of Flatcar (alpha / stable), in different Packet datacenters (Dallas, Sunnyvale, Amsterdam) and the behavior is always the same. So, while this seems to be related to the ahci kernel module, it seems to be affect a wide range of kernel versions (at least 4.19 to 5.4).

**Comment by margamanterola** [comment]
I did another test run and captured logs from the first boot, the bad boot and the re-modprobe of the ahci module. Pasting here the most relevant lines for each.

## First boot
```
[   11.678326] ahci 0000:c3:00.0: controller can do FBS, turning on CAP_FBS
[   11.698890] ahci 0000:c3:00.0: AHCI 0001.0200 32 slots 3 ports 6 Gbps 0x7 impl SATA mode
[   11.710794] ahci 0000:c3:00.0: flags: 64bit ncq fbs pio 
[   11.783773] ata10: SATA max UDMA/133 abar m2048@0xdbc00000 port 0xdbc00100 irq 264
[   11.796024] ata11: SATA max UDMA/133 abar m2048@0xdbc00000 port 0xdbc00180 irq 264
[   11.808273] ata12: SATA max UDMA/133 abar m2048@0xdbc00000 port 0xdbc00200 irq 264
[   12.126026] ata12: SATA link up 1.5 Gbps (SStatus 113 SControl 300)
[   12.144254] ata10: SATA link up 6.0 Gbps (SStatus 133 SControl 300)
[   12.247937] ata12.00: ATAPI: MARVELL VIRTUAL, 1.09, max UDMA/66
[   12.260601] ata12.00: configured for UDMA/66
[   12.284594] ata10.00: ATA-10: MTFDDAV240TCB,  D0DE012, max UDMA/133
[   12.307088] ata10.00: 468862128 sectors, multi 16: LBA48 NCQ (depth 32), AA
[   12.335413] ata10.00: configured for UDMA/133
[   12.360364] scsi 10:0:0:0: Direct-Access     ATA      MTFDDAV240TCB    E012 PQ: 0 ANSI: 5
[   12.365243] ata11: SATA link up 6.0 Gbps (SStatus 133 SControl 300)
[   12.366743] ata11.00: ATA-10: MTFDDAV240TCB,  D0DE012, max UDMA/133
[   12.366744] ata11.00: 468862128 sectors, multi 16: LBA48 NCQ (depth 32), AA
[   12.368835] ata11.00: configured for UDMA/133
[   12.371199] scsi host3: Fusion MPT SAS Host
[   12.383449] scsi 11:0:0:0: Direct-Access     ATA      MTFDDAV240TCB    E012 PQ: 0 ANSI: 5
[   12.409310] scsi: waiting for bus probes to complete ...
[   12.607951] scsi 3:0:0:0: Direct-Access     ATA      MZ7LH480HAHQ0D3  HE57 PQ: 0 ANSI: 6
[   12.607954] scsi 3:0:0:0: SATA: handle(0x000b), sas_addr(0x4433221104000000), phy(4), device_name(0x5002538e19b49869)
[   12.607955] scsi 3:0:0:0: enclosure logical id (0x54cd98f0ac7c3900), slot(0) 
[   12.607957] scsi 3:0:0:0: enclosure level(0x0001), connector name(     )
[   12.607998] scsi 3:0:0:0: atapi(n), ncq(y), asyn_notify(n), smart(y), fua(y), sw_preserve(y)
[   12.609716] scsi 3:0:1:0: Direct-Access     ATA      MZ7LH480HAHQ0D3  HE57 PQ: 0 ANSI: 6
[   12.609718] scsi 3:0:1:0: SATA: handle(0x000a), sas_addr(0x4433221100000000), phy(0), device_name(0x5002538e19b4988d)
[   12.609719] scsi 3:0:1:0: enclosure logical id (0x54cd98f0ac7c3900), slot(1) 
[   12.609720] scsi 3:0:1:0: enclosure level(0x0001), connector name(     )
[   12.609760] scsi 3:0:1:0: atapi(n), ncq(y), asyn_notify(n), smart(y), fua(y), sw_preserve(y)
[   12.611111] scsi 12:0:0:0: Processor         Marvell  Console          1.01 PQ: 0 ANSI: 5
```

## Bad boot
```
[    9.146023] SCSI subsystem initialized
[    9.357421] libata version 3.00 loaded.
[    9.516257] ahci 0000:85:00.0: version 3.0
[    9.516567] ahci 0000:85:00.0: AHCI 0001.0301 32 slots 8 ports 6 Gbps 0xff impl SATA mode
[    9.540273] ahci 0000:85:00.0: flags: 64bit ncq sntf ilck pm led clo only pmp fbs pio slum part sxs 
[    9.565058] ahci 0000:85:00.0: both AHCI_HFLAG_MULTI_MSI flag set and custom irq handler implemented
[   10.149273] ata1: SATA max UDMA/133 abar m2048@0xbf900000 port 0xbf900100 irq 121
[   10.170184] ata2: SATA max UDMA/133 abar m2048@0xbf900000 port 0xbf900180 irq 122
[   10.190618] ata3: SATA max UDMA/133 abar m2048@0xbf900000 port 0xbf900200 irq 123
[   10.210577] ata4: SATA max UDMA/133 abar m2048@0xbf900000 port 0xbf900280 irq 124
[   10.230064] ata5: SATA max UDMA/133 abar m2048@0xbf900000 port 0xbf900300 irq 125
[   10.249181] ata6: SATA max UDMA/133 abar m2048@0xbf900000 port 0xbf900380 irq 126
[   10.268058] ata7: SATA max UDMA/133 abar m2048@0xbf900000 port 0xbf900400 irq 127
[   10.286756] ata8: SATA max UDMA/133 abar m2048@0xbf900000 port 0xbf900480 irq 128
[   10.395835] ahci 0000:86:00.0: AHCI 0001.0301 32 slots 1 ports 6 Gbps 0x1 impl SATA mode
[   10.415005] ahci 0000:86:00.0: flags: 64bit ncq sntf ilck pm led clo only pmp fbs pio slum part 
[   10.543938] ata9: SATA max UDMA/133 abar m2048@0xbf800000 port 0xbf800100 irq 205
[   10.614589] ata7: SATA link down (SStatus 0 SControl 300)
[   10.643420] ata6: SATA link down (SStatus 0 SControl 300)
[   10.673354] ata5: SATA link down (SStatus 0 SControl 300)
[   10.703316] ata1: SATA link down (SStatus 0 SControl 300)
[   10.733399] ata3: SATA link down (SStatus 0 SControl 300)
[   10.762856] ata4: SATA link down (SStatus 0 SControl 300)
[   10.792166] ata2: SATA link down (SStatus 0 SControl 300)
[   10.820636] ata8: SATA link down (SStatus 0 SControl 300)
[   10.870571] ata9: SATA link down (SStatus 0 SControl 300)
[   11.398102] ahci 0000:c3:00.0: controller can do FBS, turning on CAP_FBS
[   11.426591] ahci 0000:c3:00.0: AHCI 0001.0200 32 slots 3 ports 6 Gbps 0x7 impl SATA mode
[   11.439486] ahci 0000:c3:00.0: flags: 64bit ncq fbs pio 
[   11.473058] ata10: SATA max UDMA/133 abar m2048@0xdbc00000 port 0xdbc00100 irq 264
[   11.485458] ata11: SATA max UDMA/133 abar m2048@0xdbc00000 port 0xdbc00180 irq 264
[   11.497851] ata12: SATA max UDMA/133 abar m2048@0xdbc00000 port 0xdbc00200 irq 264
[   11.814574] ata12: SATA link up 1.5 Gbps (SStatus 113 SControl 300)
[   11.836223] ata11: SATA link up 6.0 Gbps (SStatus 133 SControl 300)
[   11.938901] ata12.00: ATAPI: MARVELL VIRTUAL, 1.09, max UDMA/66
[   11.951002] ata12.00: configured for UDMA/66
[   11.976740] ata11.00: ATA-10: MTFDDAV240TCB,  D0DE012, max UDMA/133
[   11.976743] ata11.00: 468862128 sectors, multi 16: LBA48 NCQ (depth 32), AA
[   12.003616] ata11.00: configured for UDMA/133
[   12.053430] ata10: SATA link up 6.0 Gbps (SStatus 133 SControl 300)
[   12.056189] scsi host0: Fusion MPT SAS Host
[   12.072576] ata10.00: both IDENTIFYs aborted, assuming NODEV
[   12.094789] scsi: waiting for bus probes to complete ...
[   12.166262] scsi 0:0:0:0: Direct-Access     ATA      MZ7LH480HAHQ0D3  HE57 PQ: 0 ANSI: 6
[   12.195190] scsi 0:0:0:0: SATA: handle(0x000b), sas_addr(0x4433221104000000), phy(4), device_name(0x5002538e19b498)
[   12.298817] scsi 0:0:0:0: enclosure logical id (0x54cd98f0ac7c3900), slot(0) 
[   12.315952] scsi 0:0:0:0: enclosure level(0x0001), connector name(     )
[   12.332560] scsi 0:0:0:0: atapi(n), ncq(y), asyn_notify(n), smart(y), fua(y), sw_preserve(y)
[   12.352699] scsi 0:0:1:0: Direct-Access     ATA      MZ7LH480HAHQ0D3  HE57 PQ: 0 ANSI: 6
[   12.370799] scsi 0:0:1:0: SATA: handle(0x000a), sas_addr(0x4433221100000000), phy(0), device_name(0x5002538e19b498)
[   12.391666] scsi 0:0:1:0: enclosure logical id (0x54cd98f0ac7c3900), slot(1) 
[   12.409080] scsi 0:0:1:0: enclosure level(0x0001), connector name(     )
[   12.425988] scsi 0:0:1:0: atapi(n), ncq(y), asyn_notify(n), smart(y), fua(y), sw_preserve(y)
[   12.445850] scsi 11:0:0:0: Direct-Access     ATA      MTFDDAV240TCB    E012 PQ: 0 ANSI: 5
[   12.464710] scsi 12:0:0:0: Processor         Marvell  Console          1.01 PQ: 0 ANSI: 5
```

## After the `rmmod ahci` and `modprobe ahci`

```
[  306.139954] ahci 0000:85:00.0: AHCI 0001.0301 32 slots 8 ports 6 Gbps 0xff impl SATA mode
[  306.160779] ahci 0000:85:00.0: flags: 64bit ncq sntf ilck pm led clo only pmp fbs pio slum part sxs 
[  306.182730] ahci 0000:85:00.0: both AHCI_HFLAG_MULTI_MSI flag set and custom irq handler implemented
[  306.321610] ata13: SATA max UDMA/133 abar m2048@0xbf900000 port 0xbf900100 irq 121
[  306.339998] ata14: SATA max UDMA/133 abar m2048@0xbf900000 port 0xbf900180 irq 122
[  306.358156] ata15: SATA max UDMA/133 abar m2048@0xbf900000 port 0xbf900200 irq 123
[  306.376031] ata16: SATA max UDMA/133 abar m2048@0xbf900000 port 0xbf900280 irq 124
[  306.393661] ata17: SATA max UDMA/133 abar m2048@0xbf900000 port 0xbf900300 irq 125
[  306.411048] ata18: SATA max UDMA/133 abar m2048@0xbf900000 port 0xbf900380 irq 126
[  306.428171] ata19: SATA max UDMA/133 abar m2048@0xbf900000 port 0xbf900400 irq 127
[  306.445145] ata20: SATA max UDMA/133 abar m2048@0xbf900000 port 0xbf900480 irq 128
[  306.462241] ahci 0000:86:00.0: AHCI 0001.0301 32 slots 1 ports 6 Gbps 0x1 impl SATA mode
[  306.479702] ahci 0000:86:00.0: flags: 64bit ncq sntf ilck pm led clo only pmp fbs pio slum part 
[  306.510702] ata21: SATA max UDMA/133 abar m2048@0xbf800000 port 0xbf800100 irq 205
[  306.528199] ahci 0000:c3:00.0: controller can do FBS, turning on CAP_FBS
[  306.555073] ahci 0000:c3:00.0: AHCI 0001.0200 32 slots 3 ports 6 Gbps 0x7 impl SATA mode
[  306.573243] ahci 0000:c3:00.0: flags: 64bit ncq fbs pio 
[  306.627895] ata22: SATA max UDMA/133 abar m2048@0xdbc00000 port 0xdbc00100 irq 264
[  306.645401] ata23: SATA max UDMA/133 abar m2048@0xdbc00000 port 0xdbc00180 irq 264
[  306.662692] ata24: SATA max UDMA/133 abar m2048@0xdbc00000 port 0xdbc00200 irq 264
[  306.766561] ata17: SATA link down (SStatus 0 SControl 300)
[  306.781708] ata16: SATA link down (SStatus 0 SControl 300)
[  306.796649] ata13: SATA link down (SStatus 0 SControl 300)
[  306.811411] ata15: SATA link down (SStatus 0 SControl 300)
[  306.826116] ata14: SATA link down (SStatus 0 SControl 300)
[  306.838527] ata21: SATA link down (SStatus 0 SControl 300)
[  306.840372] ata20: SATA link down (SStatus 0 SControl 300)
[  306.867383] ata19: SATA link down (SStatus 0 SControl 300)
[  306.880193] ata18: SATA link down (SStatus 0 SControl 300)
[  306.990561] ata24: SATA link up 1.5 Gbps (SStatus 113 SControl 300)
[  307.003854] ata22: SATA link up 6.0 Gbps (SStatus 133 SControl 300)
[  307.107270] ata24.00: ATAPI: MARVELL VIRTUAL, 1.09, max UDMA/66
[  307.120080] ata24.00: configured for UDMA/66
[  307.130885] ata22.00: ATA-10: MTFDDAV240TCB,  D0DE012, max UDMA/133
[  307.143661] ata22.00: 468862128 sectors, multi 16: LBA48 NCQ (depth 32), AA
[  307.159093] ata22.00: configured for UDMA/133
[  307.169839] scsi 10:0:0:0: Direct-Access     ATA      MTFDDAV240TCB    E012 PQ: 0 ANSI: 5
[  307.184518] sd 10:0:0:0: [sdc] 468862128 512-byte logical blocks: (240 GB/224 GiB)
[  307.198415] sd 10:0:0:0: [sdc] 4096-byte physical blocks
[  307.209806] sd 10:0:0:0: [sdc] Write Protect is off
[  307.220837] sd 10:0:0:0: [sdc] Write cache: enabled, read cache: enabled, doesn't support DPO or FUA
[  307.225261] ata23: SATA link up 6.0 Gbps (SStatus 133 SControl 300)
[  307.238233]  sdc: sdc1 sdc2 sdc3 sdc4 sdc6 sdc7 sdc9
[  307.251055] ata23.00: ATA-10: MTFDDAV240TCB,  D0DE012, max UDMA/133
[  307.262058] sd 10:0:0:0: [sdc] Attached SCSI disk
[  307.274873] ata23.00: 468862128 sectors, multi 16: LBA48 NCQ (depth 32), AA
[  307.276957] ata23.00: configured for UDMA/133
[  307.313405] scsi 11:0:0:0: Direct-Access     ATA      MTFDDAV240TCB    E012 PQ: 0 ANSI: 5
[  307.330809] sd 11:0:0:0: [sdd] 468862128 
[  307.331106] scsi 12:0:0:0: Processor         Marvell  Console          1.01 PQ: 0 ANSI: 5
```

In particular in both the first boot and the re-modprobe we see this order of operations:

```
[   12.144254] ata10: SATA link up 6.0 Gbps (SStatus 133 SControl 300)
[   12.247937] ata12.00: ATAPI: MARVELL VIRTUAL, 1.09, max UDMA/66
[   12.260601] ata12.00: configured for UDMA/66
[   12.284594] ata10.00: ATA-10: MTFDDAV240TCB,  D0DE012, max UDMA/133
[   12.307088] ata10.00: 468862128 sectors, multi 16: LBA48 NCQ (depth 32), AA
[   12.335413] ata10.00: configured for UDMA/133
[   12.360364] scsi 10:0:0:0: Direct-Access     ATA      MTFDDAV240TCB    E012 PQ: 0 ANSI: 5
[   12.365243] ata11: SATA link up 6.0 Gbps (SStatus 133 SControl 300)
```

Whereas in the bad boot, we see this:
```
[   11.836223] ata11: SATA link up 6.0 Gbps (SStatus 133 SControl 300)
[   11.938901] ata12.00: ATAPI: MARVELL VIRTUAL, 1.09, max UDMA/66
[   11.951002] ata12.00: configured for UDMA/66
[   11.976740] ata11.00: ATA-10: MTFDDAV240TCB,  D0DE012, max UDMA/133
[   11.976743] ata11.00: 468862128 sectors, multi 16: LBA48 NCQ (depth 32), AA
[   12.003616] ata11.00: configured for UDMA/133
[   12.053430] ata10: SATA link up 6.0 Gbps (SStatus 133 SControl 300)
[   12.072576] ata10.00: both IDENTIFYs aborted, assuming NODEV
```

So, when it works, ata10 (the first drive at port 0xdbc00100) is detected as up first and it correctly configures both the ata10 and ata11 devices.  When it doesn't work, ata11 (the second drive at port 0xdbc00180) is detected as up first, and it gets configured fine, but then ata10 fails to be detected.  It very much looks like an initialization problem.  I don't know why other distros wouldn't be affected by this.

**Comment by pothos** [comment]
Thanks, sounds like we need an upstream bug report as with the crypto algboss bug.

**Comment by margamanterola** [comment]
This is definitely a bug upstream, but I think we'll need a bit more info if we want our report to get fixed.

I was thinking more about this yesterday and I realized that part of the problem (and maybe the reason of why this is Flatcar specific) is in GRUB.  What we see is that GRUB "hangs" for about 28 minutes (!? timed it twice), maybe due to a timeout or some exponential back-off retrying.   While it's stuck, GRUB shows the config options that are set. No error message, no information of what's going on, just this:

```
Welcome to GRUB!
Booting from IBA 40G Slot 4100 v1110
cmddevice=hd0
cmdpath=(hd0) Agent 40G v1.1.10
color_highlight=black/light-grayorporation
color_normal=light-gray/black
feature_200_final=y F2 1E 7E ED 00  GUID: 4C4C4544 0043 4B10 8037 C8C04F483133
feature_all_video_module=yreceived
feature_chainloader_bpb=y
feature_default_font_path=y Agent.
feature_menuentry_id=y
feature_menuentry_options=y
feature_nativedisk_cmd=y
feature_ntldr=y
feature_platform_search_hint=y
feature_timeout_style=y
grub_cpu=i386
grub_platform=pc
lang=
locale_dir=
pager=
prefix=(memdisk)
root=hd0,gpt1
secondary_locale_dir=
```

And after that, the machine boots from the initramfs, which triggers the failure to identify the root device discussed above.  But that's the same root device from which GRUB just read the initramfs !

And also, let's not forget that the first boot works just as expected. Is there something that happens during Flatcar's first boot that modifies GRUB's installation/configuration, and would cause it to behave differently upon reboot?

**Comment by pothos** [comment]
GRUB reads the boot partition contents in both cases. Yes, I also thought it would be related but now that m2 has the same problem and isn't hanging in GRUB I'm not sure – and how do we know if it's hanging in GRUB because when we see GRUB contents being the last it does not necessarily mean that it didn't start the kernel already.

**Comment by margamanterola** [comment]
Because after those 28 minutes, there's a couple of seconds when the normal GRUB screen is visible (the one that says Flatcar A and Flatcar B) and after that the boot starts with all its boot output. Also, we can see timing in the kernel logs and it matches the timing of it getting started after those 28 minutes.

**Comment by margamanterola** [comment]
Kai encountered the ahci issue but not the GRUB issue on our m2.xlarge machines (in production).  I provisioned a test m2.xlarge machine and in my case I encountered the GRUB issue, but not the ahci one... I'm not sure what this means, but just wanted to make a note of that.

After playing around with what could change between first-boot and further boots, Kai mentioned that the GPT disk UUID changes after first-boot, so I tried reverting that change, by doing: `sgdisk --disk-guid=00000000-0000-0000-0000-000000000001 /dev/sda` on my c3.medium test machine, followed by a reboot.  This caused the machine to boot correctly (both at the GRUB stage and the kernel stage). But after that boot was finished, the UUID was changed once again (I checked with `sgdisk -p /dev/sda`), and the next reboot got stuck in GRUB once again.  So, it definitely looks that something is at play with that UUID.

Meanwhile, on the m2.xlarge machine I enabled debug output in GRUB and these messages were printed after printing the config variables (and a few mallocs and frees related to parsing the options):

```
disk/i386/pc/biosdisk.c:308: Read error when probing drive 0x82
kern/disk.c:196: Opening `memdisk'...
disk/i386/pc/biosdisk.c:343: opening memdisk
lib/gpt.c:543: reading GPT from memdisk
lib/gpt.c:399: reading primary GPT from sector 0x1
lib/gpt.c:335: Primary GPT layout:
primary header = 0x7270206568742074 backup header = 0x6361622078696665
entries location = 0x20656e6f64206572 length = 0x165179adf333ed
first usable = 0x656874206f74206b last usable = 0x74636572726f6320
lib/gpt.c:445: reading backup GPT from sector 0x13
lib/gpt.c:370: Backup GPT layout:
primary header = 0x0 backup header = 0x0
entries location = 0x0 length = 0x0
first usable = 0x0 last usable = 0x0
kern/disk.c:295: Closing `memdisk'.
kern/disk.c:196: Opening `memdisk'...
disk/i386/pc/biosdisk.c:343: opening memdisk
kern/disk.c:295: Closing `memdisk'.
disk/i386/pc/biosdisk.c:282: iterating cd
kern/disk.c:196: Opening `cd'...
disk/i386/pc/biosdisk.c:343: opening cd
disk/i386/pc/biosdisk.c:457: opening cd succeeded
lib/gpt.c:543: reading GPT from cd
disk/i386/pc/biosdisk.c:598: reading 0x10 sectors at 0x0 from cd
```

And after that, it got stuck as usual with no messages of what it was doing.

**Comment by margamanterola** [comment]
More GRUB debug output, after some waiting (but still stuck). I'm capturing the messages through the out-of-band console, and it seems that it's losing some of the output, but some output is better than none. So, here's what I was able to capture:

```
disk/i386/pc/biosdisk.c:444: C/H/S 1024/255/63
disk/i386/pc/biosdisk.c:457: opening hd0 succeeded
partmap/gpt.c:90: Read a valid GPT header
partmap/gpt.c:114: GPT entry 0: start=4096, length=262144
partmap/gpt.c:114: GPT entry 1: start=266240, length=4096
partmap/gpt.c:114: GPT entry 2: start=270336, length=2097152
lib/gpt.c:543: reading GPT from hd0
lib/gpt.c:399: reading primary GPT from sector 0x1
lib/gpt.c:335: Primary GPT layout:
primary header = 0x0 backup header = 0x0
entries location = 0x0 length = 0x0
first usable = 0x0 last usable = 0x0
lib/gpt.c:445: reading backup GPT from sector 0x1bf244af
kern/disk.c:421: Read out of range: sector 0x1bf244af (attempt to read or write outside of partition).
kern/disk.c:295: Closing `hd0'.
disk/i386/pc/biosdisk.c:465: closing hd0
kern/disk.c:196: Opening `hd0,gpt2'...
disk/i386/pc/biosdisk.c:343: opening hd0
disk/i386/pc/biosdisk.c:406: LBA total = 0x1bf244b0 block size = 0x200
disk/i386/pc/biosdisk.c:444: C/H/S 1024/255/63
disk/i386/pc/biosdisk.c:457: opening hd0 succeeded
partmap/gpt.c:90: Read a valid GPT header
partmap/gpt.c:114: GPT entry 0: start=4096, length=262144
partmap/gpt.c:114: G
```

And then the same output as before:
```
disk/i386/pc/biosdisk.c:308: Read error when probing drive 0x82
kern/disk.c:196: Opening `memdisk'...
disk/i386/pc/biosdisk.c:343: opening memdisk
lib/gpt.c:543: reading GPT from memdisk
lib/gpt.c:399: reading primary GPT from sector 0x1
lib/gpt.c:335: Primary GPT layout:
primary header = 0x7270206568742074 backup header = 0x6361622078696665
entries location = 0x20656e6f64206572 length = 0x165179adf333ed
first usable = 0x656874206f74206b last usable = 0x74636572726f6320
lib/gpt.c:445: reading backup GPT from sector 0x13
lib/gpt.c:370: Backup GPT layout:
primary header = 0x0 backup header = 0x0
entries location = 0x0 length = 0x0
first usable = 0x0 last usable = 0x0
kern/disk.c:295: Closing `memdisk'.
kern/disk.c:196: Opening `memdisk'...
disk/i386/pc/biosdisk.c:343: opening memdisk
kern/disk.c:295: Closing `memdisk'.
disk/i386/pc/biosdisk.c:282: iterating cd
kern/disk.c:196: Opening `cd'...
disk/i386/pc/biosdisk.c:343: opening cd
disk/i386/pc/biosdisk.c:457: opening cd succeeded
lib/gpt.c:543: reading GPT from cd
disk/i386/pc/biosdisk.c:598: reading 0x10 sectors at 0x0 from cd
```

When it eventually booted, it didn't print anything disk or gpt related (at least not that I could capture in the OOB)

**Comment by margamanterola** [comment]
The logs above show that the entries come from "biosdisk". At first I thought this was expected, but it turns out that GRUB does have an "efidisk" component and if the machine was using EFI we would see efidisk in the logs.  Both the c3.medium and the m2.xlarge machines were booting on BIOS mode, while our c2.medium are booting on EFI mode (checked for presence of /sys/firmware/efi).

Using the OOB console I managed to set my test machines to boot UEFI mode instead of BIOS and they both booted correctly after that.

So, it seems that:
 1. Packet machines c3.medium.x86 and m2.xlarge.x86 are configured to boot BIOS mode instead of EFI mode.
 2. Our GRUB fork doesn't handle BIOS mode correctly if the disk UUID is not the default *on c3.medium and m2.xlarge machines* (as mentioned in a following comment, other machines like c1.small and c3.small don't have this issue).
 3. The ahci module also doesn't handle BIOS mode correctly if the disk UUID is not the default (?). Or maybe GRUB leaves the machine in a weird state and there's a buffer that's not correctly re-initialized until the module is unloaded and reloaded.

**Comment by pothos** [comment]
The disk UUID being randomized works without problems on many systems (that includes standard QEMU images). Not sure why it would break on those suddenly. I really wonder if the drives firmware could play a role.

**Comment by margamanterola** [comment]
Do we have anything that's booting in BIOS mode? Back when UEFI was the shiny new thing, I learned that it's not a good idea to mix GPT and BIOS mode and it might be that our forked GRUB doesn't support this situation very well.

**Comment by margamanterola** [comment]
I did a quick test with a c1.small and a c3.small machines, which are using BIOS mode to boot, and they both rebooted successfully. So, there's something more at play that includes the hardware used by c3.medium / m2.xlarge.

**Comment by pothos** [comment]
On first sight on m2.xlarge it seems to be related to the kernel version because Flatcar Stable 2247.7.0 with Linux x86_64 4.19.84-flatcar boots fine but when trying to update it doesn't find the drive in the initramfs. Maybe it's just bad luck that this only happens on the update reboot and less likely on the other reboots (and I think I hit it there too, as stated above, but I'm not sure anymore if something else was at play).

**Comment by margamanterola** [comment]
Kai identified that the issue happens during [the probing that GRUB does to find the disk with UUID 00000000-0000-0000-0000-000000000001](https://github.com/flatcar-linux/scripts/blob/flatcar-master/build_library/grub.cfg#L77).

I created a developer image that had that block commented out and tried booting it with the disk UUID having different values, and it booted fine for all values. So, everything seems to point to the fact that the call to search is probing too much. While iterating through all devices, it reaches a device that it identifies as a 'cd', and it seems to hang there (or maybe it hangs on the device after that one, but the logs do seem to point to this fake cd as the issue).

Unfortunately, while GRUB has this `--no-floppy` flag, it doesn't have a `--no-cd` flag, or any other similar flag that would prevent GRUB from probing all devices when looking of 00000000-0000-0000-0000-000000000001 (source code [[here](https://github.com/flatcar-linux/grub/blob/0863c013179c5addee289fb4037e6df39ca4e351/grub-core/commands/search_wrap.c)]).

So, given this evidence, there's two things we could do:
- Figure out why GRUB hangs on probing the device on BIOS mode but not EFI mode, fix the probing to avoid that.
- Avoid the probing altogether, by moving the UUID 0x01 detection to the initramfs instead of GRUB.

The first option is filled with too many unknowns, so we'll try to get the second one to work.

**Comment by margamanterola** [comment]
I have a fix for this, but I haven't been able to fully test it because alpha is currently broken (I've only tested it manually on the affected Packet machines, I want to do a full test run on all platforms to verify that this change doesn't break something else).  As soon as alpha gets fixed, I'll do the full test run and if everything looks good, submit the fix.

**Comment by margamanterola** [comment]
Somehow this issue didn't get updated when the fix got merged. Sorry about the lack of communication here.  The fix has been merged almost a month ago. It was first included in the alpha release 2592.0.0. It's currently only available in the alpha channel, but will move to beta with the next beta release, which should happen in a matter of days.  It should then land in stable before the end of September.

**Comment by t-lo** [comment]
Patch has arrived in Stable on Sep. 22nd, closing this.
